### DISCUSSION TITLE
An Extensive Benchmark of C and C++ Hash Tables

### DISCUSSION THREAD

**Discussion by JacksonAllan** [JacksonAllan/c_cpp_hash_tables_benchmark#1] [General]
Comments on the article [An Extensive Benchmark of C and C++ Hash Tables](https://jacksonallan.github.io/c_cpp_hash_tables_benchmark) belong below.

**Comment by Philippe91** [comment]
This is outstanding work—I understand that this type of project requires significant effort. However, here's what's missing from my point of view:

• The Qt QHash map (keep in mind that Qt is a very popular framework, so being aware of alternative maps would be beneficial). At the very least, Qt maps are used more frequently than some map implementations you evaluate.

• The benchmark focuses on large containers. In many applications, maps are not nearly as full, so data on smaller containers would have been appreciated.

**Reply by JacksonAllan** [reply]
Thanks for reading and offering feedback!

> Qt maps are used more frequently than some map implementations you evaluate.

I give some details about why I excluded some popular hash tables in [this Reddit comment](https://www.reddit.com/r/C_Programming/comments/1d3i962/comment/l67tyur/):

> Sorry, no Glib/GHashTable. This omission may be surprising, but it was based on a few considerations:
> * GHashTable only allows one type for keys and values: `void *` pointers. This severely limits its flexibility. While integers can be shoehorned into pointers (albeit with some portability concerns), larger data types need to be allocated as separate nodes. There are already two node-based or quasi-node-based tables in the benchmarks (std::unordered_map and uthash), and they—very predictably—cannot keep up with the open addressing tables due to allocation time (on insert) and pointer chasing (everywhere else). This is true even for large buckets, i.e. where some people might expect a node-based container to perform better.
> * GHashTable is, as far as I recall, now a Robin Hood hash table. There are already two Robin Hood tables in the benchmarks (ankerl::unordered_dense and tsl::robin_map), and they don't suffer from the aforementioned data-type-size constraint. In fact, there were originally three Robin Hood tables (because CC's cc_map was a Robin Hood table up until this year). I didn't see much benefit in adding another one unless there was a good reason to believe it would be more competitive than those already included (_Edit: The two remaining Robin Hood tables are both for C++. Adding Glib might be good for the sake of C users looking specifically for a Robin Hood table_).
> * I really struggled to get Glib working with MinGW (I'm on Windows), although it could be that the problem lies somewhere between the chair and the keyboard. **Generally, I was disinclined to include large libraries that do not allow their hash-table component to be installed separately, libraries that are difficult to install, libraries that would be difficult to package with the benchmarks, libraries that force the use of a particular build system, etc. (absl::flat_hash_map is the exception here, and even then it is not being included in the "proper" manner). I wanted to focus on ones that users can install and use with minimal fuss. Facebook's F14 is another rather glaring omission that I made on this basis (it explicitly does not support MinGW). Qt's QHashTable also fits into this category.** [Emphasis added]
> Please correct me if I'm wrong in any of my comments about GHashTable. It's been a long time since I looked into it. I'm not averse to adding it to the article as part of some later revision.

Another reason that some popular C++ tables don't appear is that these benchmarks started life as a benchmark suite primarily for C tables, with only the C++ tables generally regarded as the fastest and most cutting-edge included to serve as a yardstick for the C tables. However, the article has now garnered quite a bit interest from C++ developers, so I think it would be good for a revision to add (besides Glib) QHash. I'm not sure I can do much about F14, though, since including it would probably require switching to MSVC or to Linux (which would be a colossal pain for me and my one computer).

> The benchmark focuses on large containers. In many applications, maps are not nearly as full, so data on smaller containers would have been appreciated.

I touch on this issue [here](https://www.reddit.com/r/cpp/comments/1d418aw/comment/l6cyip1/) and [here](https://www.reddit.com/r/cpp/comments/1d418aw/comment/l6d2f84/). For the sake of brevity, I'll quote just the most relevant parts of those comments:

> Also, it's worth noting that the results for different key counts do not necessarily correspond to different key counts in practice. Rather, the results in the lower-key-count benchmarks should reflect the situation when the tables are hotter in the cache, whereas the high-key-count benchmarks should reflect the situation when the table are colder (I mention this point, too, in the article). In the former scenario, instructions matter most; in the latter, cache efficiency matters most. This means that tables with small key counts may behave more like what we seen in the high-key-count benchmark if they are being used infrequently or the application is doing enough other processing that there is competition over cache space (this was one of my reasons for selecting those ones for embedding with the article's text).

> I suspect the important factor might be not how many keys you have per se, but how hot you expect your table to be in the cache when the time comes to use it (e.g. if your table's keys are counted in the thousands and your program mainly does hash-table operations, look to the [0-to-200,000-key](https://jacksonallan.github.io/c_cpp_hash_tables_benchmark/result_2024-05-28T04_06_21.html) benchmarks).

I agree with the user who said that a logarithmic scale might have been better, but I also discuss some of the drawbacks of that route [here](https://github.com/JacksonAllan/Verstable/discussions/1#discussioncomment-9621565). The TLDR is that a logarithmic scale does not, in and of itself, prevent me from manipulating the impression given by the results, particularly in the heatmap, in favor of certain libraries (including my own).

**Comment by ManishJu** [comment]
thanks for the amazing work! Can you please also include 
ExcaliburHash : https://github.com/SergeyMakeev/ExcaliburHash?tab=readme-ov-file
 MeowHash : https://github.com/cmuratori/meow_hash

**Reply by wscott** [reply]
He didn't compare any hashes.

**Reply by ManishJu** [reply]
hey @wscott wouldn't some hash tables performance metrics infer some hash functions performance ? Would love to read more about these if you have some nice literature link . :)

**Reply by wscott** [reply]
I thought both the links you posted were hash functions and not hash tables, but I see now that the first is a hash table and only the second is a hash function.

from the article:
> All the tables benchmarked are configured to use the same hash functions, namely the [MurmurHash3 finalizer](https://zimbry.blogspot.com/2011/09/better-bit-mixing-improving-on.html) for integer keys and [FNV-1a](https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function) for string keys.

So he didn't compare how changing the hash function would impact different data structures.

If you want to look at how changing the hash function impacts performance look [here](https://martin.ankerl.com/2022/08/27/hashmap-bench-01/) which was also linked in the article.

**Comment by bftjoe** [comment]
Please add PGM index. https://github.com/gvinciguerra/PGM-index/blob/master/examples/updates.cpp

PGM index doesn't even need a hash function for integer keys.

And maybe fph https://github.com/renzibei/fph-table

**Comment by nlescoua** [comment]
Great article, for C you might be interested in :
https://github.com/technicolor-research/cuckoopp

It features a cuckoo hash table with memory layout optimized for SIMD lookup, and the addition of a small bloom filter in each bucket to improve negative lookups. It can work looking up single value but also offers a batch lookup interface that allow further performance optimizations like prefetching, loop unrolling,…

**Comment by Zhihan-Lu** [comment]
Amazing work! Thanks a lot for the detailed explanation.

Does the author have statistics for the CPU and Memory usage when benchmarking each table? I am wondering when the system's CPU load is high, whether the performance of the SIMD tables will be impacted.

**Reply by JacksonAllan** [reply]
Thanks for reading!

I talk a little about the fact that the benchmarks do not measure memory usage [here](https://old.reddit.com/r/cpp/comments/1d418aw/an_extensive_benchmark_of_c_and_c_hash_tables/l6clh5e/). The relevant part of my comment reads:

> The fact that these benchmarks do not measure memory usage is probably their biggest weakness. I made this choice early on in order to limit the scope of the project, but I ended up regretting it as I added more tables that displayed more varied ways of storing data and therefore couldn't be easily compared—in terms of memory usage—to the others based only on one or two numbers (e.g. per-bucket-overhead). Ultimately, I decided that if I'm going to measure memory usage, it will have to be in some kind of follow-up article/project. But do take a look at [this benchmark](https://github.com/attractivechaos/udb3), which—although it only includes one pair of data types—does measure memory usage and look at speed in the context of it. *Edit: Also, I believe that martinus' [benchmarks](https://martin.ankerl.com/2022/08/27/hashmap-bench-01) included memory usage.*

Unfortunately, integrating measurements of memory usage into the project would require serious restructuring so that each table is benchmarked in a separate process rather than in one monolithic benchmarking program. But for most of the open addressing tables, it should hopefully be possible to make a fairly accurate estimate of their memory usage based on the description I included in the article, especially because they mostly use the same growth policy and therefore grown in sync with one another.

> I am wondering when the system's CPU load is high, whether the performance of the SIMD tables will be impacted.

That's a good question. I'll try to summon @joaquintides, who developed boost::unordered_flat_map, to see whether he can shed any light on the matter. My own experiments with SIMD table were less successful than the Boost table, so I moved away from SIMD in favor of in-table chaining early in the development of my libraries (although that's I choice I might have reconsidered if I'd known about Boost's performance at that time).

**Reply by joaquintides** [reply]
> > I am wondering when the system's CPU load is high, whether the performance of the SIMD tables will be impacted.

> That's a good question. I'll try to summon @joaquintides, who developed boost::unordered_flat_map, to see whether he can shed any light on the matter. 

I don't have any significant insight to provide here. The tables tested are single threaded, so an increase in CPU load from external threads/processes boils down to having time slicing for the table-related code and to more cache pressure. This would result in degraded performance, maybe in a higher degree for the more cache-friendly tables, but I'm speculating here.

**Reply by tnodir** [reply]
Can you please also add [TommyDS](http://www.tommyds.it)?

**Comment by MaskRay** [comment]
It might be nice to have more analysis on memory consumption and code size.

Do you have recommendation for improvements on https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/DenseMap.h ? It  is like Google's original dense hash map. It has strong constraints on code size and memory consumption, which might make migration to other hash maps difficult.
Its load factor 75% is low, though.

**Reply by aws-taylor** [reply]
+1 It would be awesome if the LLVM data structures were included in the analysis.

**Comment by nmehran** [comment]
This project has been a source of inspiration in some internal testing I have done -- thank you for your time and effort.   

I want to share a folly environment you can develop and benchmark with, on any platform, using Docker: [folly-docker-env](https://github.com/nmehran/folly-docker-env) 

I have performed benchmarks for close to 90% of the libraries requested above, and will open-source the code and results in a matter of time.

**Reply by JacksonAllan** [reply]
> I have performed benchmarks for close to 90% of the libraries requested above, and will open-source the code and results in a matter of time.

Great! I'm happy someone is getting some use out of the project and very interested in seeing your results. Unfortunately, so far I've done a poor job of expanding the benchmarks with more libraries myself, as is evident from the previous unaddressed comments from other users.

Also, @gaujay had the idea of adding an [extra benchmark](https://old.reddit.com/r/cpp/comments/1g2oso8/introducing_flat_umap_a_fast_simdbased_unordered/mjv4jtx/) to test the performance of reinsertion after erasure. I've been meaning to incorporate it (especially as my own table should, I think, excel in that area thanks to its "true" erasures).

**Reply by nmehran** [reply]
You've contributed a great deal -- thank you for that.  This work interests me as well, for various reasons.  I see no issue with creating the additional benchmark, and I’ll share the shims mentioned above once complete, along with a few bonuses.  Sit back and relax for this iteration :)

**Reply by gaujay** [reply]
Yes, thanks again @JacksonAllan! You've made a lot of things possible thanks to your benchmark.

I see the Godbolt link is still up with the snippet for that extra use case: https://godbolt.org/z/TK6MMYPT5
The idea was to showcase the effect of tombstones by erasing all but one element (to avoid hidden clear mechanism), then re-adding them.

I also added some other hash tables at the time (like Folly) to benchmark my own, with the following results:
https://github.com/gaujay/indivi_collection/tree/main/bench/flat_unordered
If you think it's worth it, I'd really appreciate if _flat_umap_ could be added to the race :)

**Reply by nmehran** [reply]
I've completed all hash-table implementations. Libraries like Folly, LLVM, and Qt were especially difficult to integrate -- they'll require a containerized build, which I’ll publish.

Cuckoo++ (cuckoopp) was a unique challenge. It depends on DPDK and Ubuntu 16.x (GCC 5.x), so I built a dedicated container. However, due to poor benchmark results and its fixed-size nature (no rehashing), I’ve pruned that branch. Still, it showed some interesting behavior and good performance with 16-byte c-strings.

I’ll release a minimally intrusive patch with the new shims and benchmark. Libraries with heavy dependencies will require containerized execution.

One important finding: c_cpp_hash_tables_benchmark shows order-dependency bias. Despite the `flush_cache` logic, results vary significantly based on shim order. The leading hypothesis is that CPU caches, allocator state, or branch predictors aren’t fully reset between runs, causing cross-shim contamination. I’ve implemented a fix, but it’s intrusive and will require a significant refactor of the codebase.

P.S. Would you prefer the shims and benchmark updates shared as two separate pull requests, a single pull request, or a fork?

With regard to the order-dependency bias, see the discrepancy in "Insert nonexisting" performance below, based on the run-order:

![Screenshot from 2025-06-05 00-54-31](https://github.com/user-attachments/assets/003853b9-8226-4e29-bb8a-b22a645a2f78)

![Screenshot from 2025-06-05 00-54-39](https://github.com/user-attachments/assets/0222d2dd-db2b-474f-9101-f92390bfdf83)

**Reply by gaujay** [reply]
Back then, I experienced some noise between runs too but not as significant as this.
I tried switching benchmarks order a few times but didn't notice much difference.
What kind of CPU did you run on? Did you disable its frequency scaling beforehand?

**Reply by JacksonAllan** [reply]
> I've completed all hash-table implementations.

Fantastic! Amazing work.

> One important finding: c_cpp_hash_tables_benchmark shows order-dependency bias. Despite the flush_cache logic, results vary significantly based on shim order. The leading hypothesis is that CPU caches, allocator state, or branch predictors aren’t fully reset between runs, causing cross-shim contamination. I’ve implemented a fix, but it’s intrusive and will require a significant refactor of the codebase.

Interesting. I toyed around with randomizing the order at one point during development but didn’t notice much difference. However, I was at that time focusing on the largest benchmarks – the 20,000,000-key set – which might be the most stable in this regard. The final order in the original project was language (C++ and then C) and then alphabetical based on the name of the library (to show that I wasn't using order to manipulate the results).

How did you implement your fix? The best solution might be to separate each table into its own dedicated program that outputs its results into a text/CSV file, run each program in succession, and then run another program that reads all the results and generates the graphs and tables. This whole process could be automated with a makefile that builds all the programs and generates the necessary script (e.g. Bash/PowerShell) to run them all and generate the graphics. The idea of separating each table into its own program (which is honestly what I should have done from the beginning) came up during the [discussions](https://github.com/JacksonAllan/Verstable/discussions/1) that went on behind the scenes during the project’s development (albeit for the purpose of measuring memory consumption, which is a major gap in the original project and article). Naturally, this would be a major refractor. I could have a go at it, but if this is what you already did, I’d be repeating your work.

This might be a good time for me to mention that it would also be good to combine each graphic for the three different key-counts into one graphic where the key count can be toggled. This would help compensate for the fact that the scale is linear, rather than logarithmic. Of course, such a combined graphic wouldn’t work when displayed in a GitHub README, where interactivity is disabled, but that’s probably not very important.

> P.S. Would you prefer the shims and benchmark updates shared as two separate pull requests, a single pull request, or a fork?

Let me have a think about that. Ideally, the original code should remain available for my 2024 article to continue linking to. I’m leaning towards the idea of keeping everything together in the same repository but having a v1 branch with the original code and v2 (i.e. main) branch with these improvements (Maybe I can add you as a collaborator/maintainer? I’m not too sure how that works.). However, it seems like you’ve put a lot of work into this, so maybe you’d prefer to maintain your own public fork?

By the way, are you planning to share your results anywhere? I think some of the libraries you added – e.g. Qt – were requested in the Reddit comments on the original article.

**Reply by nmehran** [reply]
Yes, I appreciate your input @gaujay 

I run benchmarks extensively, and can share a few of the scripts I use:
- manage CPU frequency scaling and boost: [manage_cpu_scaling.sh](https://gist.githubusercontent.com/nmehran/1ec4c117356809cc0a778de5e84a3a28/raw/7c569e225757b591eab517801fafb2279c82869f/manage_cpu_scaling.sh)
- manage CPU affinity and core-pinning: [cpu_topology.h](https://gist.githubusercontent.com/nmehran/1ec4c117356809cc0a778de5e84a3a28/raw/7c569e225757b591eab517801fafb2279c82869f/cpu_topology.h)

See [test hardware](https://pastebin.com/NEAexaCu).

I want to clarify, the variance seen in the demonstration would be more subtle under the normal 200,000 key benchmark:

![Screenshot from 2025-06-05 19-37-31](https://github.com/user-attachments/assets/f450533f-0d3e-426d-8503-486e17cb9cec)
![Screenshot from 2025-06-05 19-37-23](https://github.com/user-attachments/assets/00e9b914-6850-4239-9dc7-bd1970efd288)

The results from my internal suite seem more consistent with Verstable's design (at keys: 200,000):

![image](https://github.com/user-attachments/assets/dd352174-11b5-4ba1-ae6e-4174a6bbe2b3)

@JacksonAllan I'll go into detail soon, but in short: stability comes from more robust flushing, per-repetition order-cycling, better key handling, and eliminating other runtime biases -- and though I have a version which supports multiple processes, it's not required to get stable results.  As a hash-table developer, I know how crucial stability is -- we need to measure changes at the nanosecond level to measure improvements.

For the scope of this contribution, I will create a minimally intrusive fork (no other changes other than the benchmark and shims), and we can merge (or not) at your discretion.  For the stability changes, I will create a branch you can inspect, as it may differ significantly (I run benchmarks on a different internal codebase, so this one will be adapted).

I'll likely do a final run using the containerized version (with Qt, etc) once the code is released and publish the HTML results on GitHub Pages, but it will be easy for anyone to run via Docker and reproduce with c_cpp_hash_tables_benchmark, given the shims and container.

**Reply by gaujay** [reply]
Looking back at the benchmark code, some differences between runs/ordering might be due to the random generator usage.

We are using `std::default_random_engine` and `std::uniform_int_distribution` which are not portable between platforms/compilers.
A better choice could be `std::mt19937` (Mersenne Twister) for example.

We are also initializing the engine as a global with `steady_clock::now`, making its output different for each process run:
- `keys` vector is init as static, making it identical for all shims, but changing its order between process runs
- most operations (get/erase/replace/iterate) are random-based too, making them different for each shim/bench run

Maybe I missed something reading the code, but if not, it would be better for repeatability/stability to pre-compute all random-based input values and re-use them between bench runs/shims.
Finding and hard-coding a "good" seed for the random engine could make results more stables between executions too
(i.e. without particular bias compared to others seeds).

**Reply by nmehran** [reply]
Determinism is key.  Agreed @gaujay you are spot on with what you've pinpointed, among other things.

I was still seeing fluctuations, and there was a major bug in the internal suite’s definition of "reinsert after erasure," which differed from the one I used in c_cpp_hash_tables_benchmark. I’ve fixed it, and now variance between runs is under 3% --essentially the same result every time. This discrepancy was a key reason I held off on releasing. The "true" results make much more sense:
 
![Screenshot from 2025-06-07 19-09-52](https://github.com/user-attachments/assets/1477128e-2d56-4132-8a0a-8507eb513b91)

Also, system-level stability hardening is another layer (disabling SMT, ASLR, setting OS-level core management).

**Reply by JacksonAllan** [reply]
> We are using `std::default_random_engine` and `std::uniform_int_distribution` which are not portable between platforms/compilers. A better choice could be `std::mt19937` (Mersenne Twister) for example.
> 
> ...
> Maybe I missed something reading the code, but if not, it would be better for repeatability/stability to pre-compute all random-based input values and re-use them between bench runs/shims.

Could we not just re-seed Mersenne Twister (or one of the faster, simpler PRNGs that have emerged since mt19937 was introduced into the Standard — I don't think we need super high quality random numbers here) with the same seed before each shim/benchmark or run? I think that might be better than precomputing a large array of random values, which could introduce another opportunity for cache misses unrelated to the hash table being benchmarked to influence the results.

> I was still seeing fluctuations, and there was a major bug in the internal suite’s definition of "reinsert after erasure," which differed from the one I used in c_cpp_hash_tables_benchmark. I’ve fixed it, and now variance between runs is under 3% --essentially the same result every time.

Fantastic, good to know.

Also, allow me to tag the usual suspects who will probably be interested in the results of the improved benchmarks once they're available: @joaquintides, @attractivechaos, and @martinus. These are the developers of boost::unordered_flat_map, ankerl::unordered_dense, and khash/khashl, respectively. Each one offered input into the original project (though the shortcomings in it are my own!) and has published his own sets of hash-table benchmarks.

**Reply by martinus** [reply]
In nanobench https://github.com/martinus/nanobench I have an extremely fast RNG that's I usually use for benchmarking. See documentation here: https://nanobench.ankerl.com/reference.html#_CPPv4N6ankerl9nanobench3RngE
The quality of the RNG is not the best, but it is very fast and more than good enough for benchmarking.

**Reply by joaquintides** [reply]
I can't seem to find the source code for the benchmarks including `indivi_flat_umap`. FWIW, I tested this container against Boost and Ankerl some months ago, results and code here:
https://github.com/boostorg/boost_unordered_benchmarks/tree/boost_unordered_aggregate_indivi

**Reply by nmehran** [reply]
@martinus Nanobench looks great.. lots of useful features, and your code style is a pleasure to read -- thank you.

@joaquintides Thanks for your work on Boost.Unordered -- I’d call it *the* current benchmark for open-source hash tables, and excels at everything that matters for an LRU cache.

> another opportunity for cache misses unrelated to the hash table being benchmarked

@JacksonAllan Loving where this discussion is headed.

I believe that as long as the bias is deterministic and fair, relative performance (against itself prior to a change, or another hash table) is enough for constructive feedback in evaluating micro-optimizations, regardless of noise from cache-misses in key retrieval, etc.  The raw timings definitely are not as accurate as they can be, though I don't think that's too important, as long as "relative" evaluation is done in the same setting.

---

I'll publish fully as I review and polish, but at a minimum, here is [the new benchmark and indivi](https://github.com/nmehran/c_cpp_hash_tables_benchmark/tree/shims) ( well done @gaujay ), with no other changes.  @JacksonAllan @gaujay Let me know if I overlooked anything with regard to the benchmark.

Thanks again for your contribution @JacksonAllan

![Screenshot from 2025-06-09 15-29-35](https://github.com/user-attachments/assets/87c2bb63-2efa-4c3f-9dd0-28ac0c40cdf8)

**Reply by nmehran** [reply]
The [shims branch](https://github.com/nmehran/c_cpp_hash_tables_benchmark/tree/shims) is complete. Feel free to use them in your own projects.

* `cwisstable`
* `excalibur`
* `folly_f14_fast`
* `folly_f14_node`
* `folly_f14_value`
* `folly_f14_vector`
* `fph_dynamic_map`
* `indivi_flat_umap`
* `llvm_dense_map`
* `pgm_dynamic`
* `tommyds_dyn`
* `tommyds_hash`
* `tommyds_lin`
* `qt6_hash`

Use [this container and branch](https://github.com/nmehran/hash-env-docker/tree/c_cpp_hash_tables_benchmark/main) to run Qt, LLVM, and Folly with `c_cpp_hash_tables_benchmark`. It works on Linux, macOS, and Windows, and includes scripts for scaling, huge pages, and live-mounting local code (or config) without rebuilding the container.

A minimal branch remains to demonstrate result stabilization techniques and share a few small surprises.

**Reply by gaujay** [reply]
Thanks @nmehran, did you have the chance to make progress on an updated table with all results?
I think lots a people would be interested to see a broader comparison with all these hash maps (including those small surprises).

**Reply by nmehran** [reply]
I've been quite busy, but I appreciate the follow up @gaujay @JacksonAllan.  

It is quite easy to run the updated benchmark via the command `./run_hash_env.sh ./run_benchmarks.sh` via [hash_env_docker](https://github.com/nmehran/hash-env-docker) once you check out the `c_cpp_hash_tables_benchmark/main` branch in that repo.  It includes the new "reinsert after erasure" and all of the shims (you just need to select them in the config.h file, as before).

I have ran the benchmarks on 20,000 - 20,000,000 and am happy to share the results.  I've also decoupled the html and csv output logic from the main.cpp into a `benchmark_output.h` file and enhanced the heatmap and graphs.  The new output code is open source in the dev branch [here](https://github.com/nmehran/c_cpp_hash_tables_benchmark/tree/dev) and most results are here: [https://nmehran.github.io/hash-table-benchmark-report/](https://nmehran.github.io/hash-table-benchmark-report/)

I had to make some choices as to what to include in this report.  I've included top performers, where a hash table led in at least one category, classics (abseil), and popular demand (llvm, qt, folly).  Others that were omitted were simply too slow (noncompetitive), or unfavorable (could not change the hash function, could not resize, etc).

**Reply by gaujay** [reply]
Looks great! The new interactive options are really nice, thanks.
Was it run on a Linux/GCC x64 machine, using murmur/fnv hashes?
(Note: you might want to add some cache_flush after insertions to reduce bias/noise)

The lib 'omm' seems to perform well but I don't see a shim for it in the repo.
Looking at graphs, I'm guessing it's a variation on Boost design (maybe without the anti-drift mechanism?)

**Reply by nmehran** [reply]
Thanks @gaujay, and you're welcome!

Yes, this was run on a standard Linux/GCC x64 machine, using the standard murmur/fnv hashes and benchmark definitions in this repo. The main difference is that I used a more advanced internal benchmark suite that employs several techniques to reduce noise and bias. This includes using a dedicated dispatcher core, cycling the test order for each run, implementing more robust cache flushing (including where you mentioned), and other tricks. It uses a task queuing system. I plan on adding some of these enhancements to the public dev branch when I get the time!

The omm library is a privately licensed C library for now, but it may become open-source in the future.

**Comment by oliness** [comment]
I tried it with boost 1.88 (current version). Boost is slightly better on my 1.8ghz i7 with 1.88, although it's still weaker at Iterating than anything else.

I tested by adding the boost path to a g++ include, do you have a one-liner to copy it into a single file like for 1.85? Would be good to keep it up to date.

**Comment by gaujay** [comment]
Finally published an updated version of the suite:
https://github.com/gaujay/c_cpp_hash_tables_benchmark

I added the 're-insert after erase' benchmark (to showcase tombstones effect)
along with some minor changes for better cross-platform repeatability.

I did it to test a new unordered map I was working on (flat_wmap).
Updated benchmark results with different compilers are available here:
https://github.com/gaujay/indivi_collection/tree/main/bench/flat_unordered

**Reply by oliness** [reply]
Looks great, would you accept a pr for changing the Boost test to use a checkout of boost_unordered?

Currently at 1.88, can then be quickly updated with new versions to track improvement/regression. Boost also have a number of related containers and I'd like to test them all.

**Reply by gaujay** [reply]
I like the idea but not sure this repo is the right place, as it still misses lots of things:
- not all shims support the new benchmark (some miss the new 'eraseN()' method)
- added the shims for folly and indivi, but not the libs
- not all libs support the new std::string data type

It would be nice however to incorporate some of the improvements to the original repo, including your proposition.
PS: I did some tests with boost 1.88 but I'm not sure there was any performance change since 1.84

**Reply by JacksonAllan** [reply]
Apologies for being absent for a while - I've been inundated with work. I've set aside a day mid-week to go through all the new developments and hopefully update the original repo. For now, let me just thank everyone (@gaujay, @oliness, @nmehran, @joaquintides, @martinus) for taking the time to help improve this project.

**Reply by gaujay** [reply]
Great! Just updated my fork again to remove the need for a new shim function (eraseN) to enable the re-insert benchmark.
Also added indivi lib as an embedded third-party for convenience (separate commit).

If you're interested, here are some notes about the other changes:
- Added some flush_cache calls between insert and other operations to avoid cache hotness bias.
Now the performance for data fitting in L1 cache (i.e. small maps) is properly growing with size.
- Changed default operation count from 1000 to 500. This allow to keep curves details better (with less caching bias, for small maps especially) while keeping noise low. I also fixed the iteration count that was 1000 instead of the expected 5000.
- Replaced random engine/distribution with portable ones (std::mt19937_64 and modulo). Also use fixed seeds per run, so every benchmark use the same pseudo-random data.
- Changed blueprints integer keys from 0-N sequences to random values, also to avoid bias.

**Comment by lefty2** [comment]
I tried boost::unordered_flat_map with the pathfinding algorithm for my game and it was 30% slower than regular boost::unordered_map. No idea why. Maybe unordered_flat_map is only fast for large amounts of keys?

**Reply by joaquintides** [reply]
Hi @lefty2, author here. Can you share a little more context, like which operations are faster, number of elements, the definition of the maps you're using (including the hash function) etc.? In principle, `boost::unordered_flat_map` should be faster than `boost::unordered_map`, but maybe you've hit some special scenario.

**Reply by lefty2** [reply]
when I tested it was using around 90 elements in the map. The key is an int and the value is a pointer. So the algorithm is doing some find, followed by some insertions. The it iterates through then entire map and does one erase and continues util the path is found:
```cpp
	boost::unordered_flat_map<int, PathTile*>::iterator iter;
	boost::unordered_flat_map<int, PathTile*>::iterator iterLowest;

    Point2D posDest, posCurr;
	PathTile* pPathTile;
	PathTile* pCurrTile = m_pStartPath;
	do
	{
		posCurr = pCurrTile->GetPosGrid();
		for (int n = 1; n < 9; n++)
		{
			g_surroundCell.GetSurround(n, posCurr, posDest);
			iter = m_mapClosed.find(posDest.GetKey());
			if (iter != m_mapClosed.end()) continue;
			int doorChange = pCurrTile->GetDoorChange();
			bool ignoreCenterColl = endPointTarget && posDest == end;
			if (m_pSceneGrid->CanMoveTo(pCombatant, posCurr, posDest, ignoreCenterColl, doorChange))
			{
				iter = m_mapOpen.find(posDest.GetHashKey());
				if (iter != m_mapOpen.end())
				{
					iter->second->CheckBetterParent(pCurrTile, end, doorChange);
					continue;
				}
				else
				{
					pPathTile = m_listCache.AddNew();
					pPathTile->Init(pCurrTile, posDest, end, doorChange);
					m_mapOpen[posDest.GetHashKey()] = pPathTile;
				}
			}
		}
		if (m_mapOpen.size() == 0 )
		{
			return;
		}
		// find lowest
		iterLowest = m_mapOpen.begin();
		iter = m_mapOpen.begin();
		iter++;
		for (;iter != m_mapOpen.end(); iter++)
		{
			if (iter->second->IsBetter(iterLowest->second))
				iterLowest = iter;
		}
		pCurrTile = iterLowest->second;
		m_mapClosed[iterLowest->first] = pCurrTile;
		m_mapOpen.erase(iterLowest);
	} while (pCurrTile->GetPosGrid() != end);
```

**Reply by joaquintides** [reply]
Ok, so you have two maps `m_mapOpen` and `m_mapClosed`, with keys obtained from `GetHashKey` and `GetKey`, respectively. What do these keys look like? Are they more or less random numbers or sequential ones?

**Reply by lefty2** [reply]
The key is the y co-ordinate shifted 16 + the x co-ordinate (y << 16 | x)

**Reply by joaquintides** [reply]
I don't see anything strange there... well, I don't know, it's possible that for containers this small `boost::unordered_map` is a better option.

**Reply by lefty2** [reply]
thanks for the answer anyways

**Reply by joaquintides** [reply]
On an unrelated note, you can probably speed up this part:

```cpp
iter = m_mapOpen.find(posDest.GetHashKey());
if (iter != m_mapOpen.end())
{
    iter->second->CheckBetterParent(pCurrTile, end, doorChange);
    continue;
}
else
{
    pPathTile = m_listCache.AddNew();
    pPathTile->Init(pCurrTile, posDest, end, doorChange);
    m_mapOpen[posDest.GetHashKey()] = pPathTile;
}
```
like this:
```cpp
template<typename F>
struct deferred_call
{
    using result_type = decltype(std::declval<const F>()());
    operator result_type() const { return f(); }

    F f;
};
...

auto [iter, b] = m_mapOpen.try_emplace(
    posDest.GetHashKey(),
    deferred_call([&] {
        pPathTile = m_listCache.AddNew();
        pPathTile->Init(pCurrTile, posDest, end, doorChange);
        return pPathTile;
    }));
if (!b) {
    iter->second->CheckBetterParent(pCurrTile, end, doorChange);
}
```
More details [here](https://bannalia.blogspot.com/2022/10/deferred-argument-evaluation.html).
